cluster-api/apis/cluster/types.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package cluster contains types to represent Kubernetes cluster configuration.
+package cluster // import "k8s.io/kube-deploy/cluster-api/apis/cluster"
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// Cluster is an API object representing a cluster's control-plane
+// configuration and status.
+type Cluster struct {
+	metav1.ObjectMeta
+	metav1.TypeMeta
+
+	Spec   ClusterSpec
+	Status ClusterStatus
+}
+
+type ClusterSpec struct {
+	// Nominal version of Kubernetes control-plane to run.
+	KubernetesVersion KubernetesVersionInfo
+
+	// Basic API configuration
+	APIConfig APIServerConfig
+
+	// Cluster network configuration
+	ClusterNetwork ClusterNetworkingConfig
+
+	// Provider-specific serialized configuration to use during
+	// cluster creation. It is recommended that providers maintain
+	// their own versioned API types that should be
+	// serialized/deserialized from this field.
+	ProviderConfig string
+}
+
+type APIServerConfig struct {
+	// The address for the API server to advertise.
+	AdvertiseAddress string
```

Comment:
I'm wondering about the semantics of `AdvertiseAddress`. It makes a lot of sense in the `kubeadm init` case where I already have a VM provisioned and am just initializing it, but how should I use it here? Should I omit this at cluster creation time, and let the installer + cloud provider assign it automatically when it brings up my master node? Is it meant just to differentiate between internal network interface binding vs. external? Or is it to support tooling that needs to modify master hosts, and is meant to help direct traffic around while we take masters out of service and replace them with new ones, without causing an outage?

Reply:
After reading further, I assume that this field is specifically to configure the apiserver, and that `APIEndpoint.Host` is the endpoint for other controllers / clients to use. In that case, I would expect `APIEndpoint.Host` to potentially be a load-balancer, but that `APIConfig.AdvertiseAddress` would be a single address for an individual host. How does that work with mult-masters? Am I completely misunderstanding this field?

Reply:
You are not. I've been thinking about how to actually approach this clash of shared and specific config in the control plane. I've been telling myself that we'll just setup one master for the time being and solve it later, but we need address at least the representation now.

I think we need a representation that enforces cluster-wide config for things that have to be the same across the entire control plane (like pod and service CIDRs), but also support individual controller attributes that don't have to agree (like perhaps advertising addresses, certificate information...). 

I think a natural place for the individual controller attributes to live would be an optional field in the machine API that should only be set when the role for the machine is a master. What do you think? We could keep this top-level cluster object specify cluster-wide config and cluster-wide status. This would also give us a natural way to expand HA and do individual master upgrades.